### PR TITLE
HIVE-28446: Convert some reserved words to non-reserved words

### PR DESCRIPTION
**Pull request by okumin**
### What changes were proposed in this pull request?

Mark some reserved words as non-reserved based on the ANSI standard. I did the following.

1. [Write unit tests using the existing parser](https://github.com/apache/hive/commit/40a2b31c9eeb602d773483bf9bab5581794c4799)
2. Pick up keywords that are not reserved in the ANSI standard
    - https://www.postgresql.org/docs/9.5/sql-keywords-appendix.html
    - SQL:2023
3. Add reserved words which are not declared [in the WIKI](https://cwiki.apache.org/confluence/display/hive/languagemanual+ddl#LanguageManualDDL-Keywords,Non-reservedKeywordsandReservedKeywords) to the non-reserved list

### Why are the changes needed?

We found some extra reserved keywords have been added while we were testing Hive 2 -> Hive 4. I guess some of them are intentionally reserved, but some of them are unintentionally reserved.

One point. We're not biased toward making everything non-reserved. It is OK to discard this PR and add all keywords to [the WIKI](https://cwiki.apache.org/confluence/display/hive/languagemanual+ddl#LanguageManualDDL-Keywords,Non-reservedKeywordsandReservedKeywords).

### Does this PR introduce _any_ user-facing change?

Users can use the changed words without adding quotes.

This PR would not tighten any restrictions, so I assume this is backward compatible.

### Is the change a dependency upgrade?

No.

### How was this patch tested?

Added unit tests, and checked the behavior of Trino and PostgreSQL.

```
trino> create table hive.default.test_keywords (APPLICATION int, COMPACTIONID int, CONF int, CONNECTOR int, CONNECTORS int, DATABASE int, DDL int, EXCHANGE int, EXTENDED int, FOLLOWING int, FORCE int, IF int, LESS int, MACRO int, MORE int, OLDER int, PKFK_JOIN int, PRECEDING int, SYNC int, THAN int, TIMESTAMPLOCALTZ int, UNBOUNDED int);
CREATE TABLE
```

```
postgres=# create table test_keywords (APPLICATION int, COMPACTIONID int, CONF int, CONNECTOR int, CONNECTORS int, DATABASE int, DDL int, EXCHANGE int, EXTENDED int, FOLLOWING int, FORCE int, IF int, LESS int, MACRO int, MORE int, OLDER int, PKFK_JOIN int, PRECEDING int, SYNC int, THAN int, TIMESTAMPLOCALTZ int, UNBOUNDED int);
CREATE TABLE
```

### CODE REVIEW

File: parser/src/java/org/apache/hadoop/hive/ql/parse/IdentifiersParser.g [hunk 1]
```diff
@@ -988,10 +988,32 @@ nonReserved
...
+    | KW_MACRO
+    | KW_MORE
+    | KW_OLDER
+    | KW_PKFK_JOIN
+    | KW_PRECEDING
+    | KW_SYNC
+    | KW_THAN
+    | KW_TIMESTAMPLOCALTZ
+    | KW_UNBOUNDED
 ;
```
Comment (okumin) [inline parser/src/java/org/apache/hadoop/hive/ql/parse/IdentifiersParser.g:1013]: The current commit maximized the number of non-reserved words. But I think some of them(or all?) are intentionally configured. I guess we would like to revert some of them.
https://cwiki.apache.org/confluence/display/hive/languagemanual+ddl#LanguageManualDDL-Keywords,Non-reservedKeywordsandReservedKeywords

Just for convenience,

- Introduced by 1.2.0
    - CONF, DATABASE, EXCHANGE, EXTENDED, FOLLOWING, IF, LESS, MACRO, MORE, PRECEDING, UNBOUNDED
- Introduced by 3.0.0
    - SYNC
- Not mentioned in the document
    - APPLICATION, COMPACTIONID, CONNECTOR, CONNECTORS, DDL, FORCE, OLDER, PKFK_JOIN, THAN, TIMESTAMPLOCALTZ

File: parser/src/java/org/apache/hadoop/hive/ql/parse/IdentifiersParser.g [hunk 2]
```diff
@@ -988,6 +988,16 @@ nonReserved
...
+    | KW_COMPACT_ID
+    | KW_DATACONNECTOR
+    | KW_DATACONNECTORS
+    | KW_DDL
+    | KW_FORCE
+    | KW_OLDER
+    | KW_PKFK_JOIN
+    | KW_THAN
+    | KW_TIMESTAMPLOCALTZ
 ;
```
Comment (okumin) [inline parser/src/java/org/apache/hadoop/hive/ql/parse/IdentifiersParser.g:1001]: Tickets and fix versions

- KW_APPLICATION: HIVE-18004(fix version=3.0.0)
- KW_COMPACT_ID: HIVE-26580(4.0.0)
- KW_DATACONNECTOR  & KW_DATACONNECTORS: HIVE-24396(4.0.0)
- KW_DDL: HIVE-24596(4.0.0)
- KW_FORCE: HIVE-4367(4.0.0)
- KW_OLDER & KW_THAN: HIVE-27906(4.0.0)
- KW_PKFK_JOIN: HIVE-23817(maybe 4.0.0)
- KW_TIMESTAMPLOCALTZ: HIVE-16614(3.0.0)
    - [This could have replaced KW_TIMESTAMPTZ](https://github.com/apache/hive/commit/34eebff194e81180202d198200e84058c4910d95#diff-0873af31a12d73e96355bffe6b8a1435c958c744231f0ccd3f80d057920cb8ce)

File: parser/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g [hunk 3]
```diff
@@ -2438,7 +2438,6 @@ primitiveType
     | KW_DATETIME      ->    TOK_DATETIME
     | KW_TIMESTAMP     ->    TOK_TIMESTAMP
     | KW_TIMESTAMPLOCALTZ   ->    TOK_TIMESTAMPLOCALTZ
-    //| KW_TIMESTAMPTZ   ->    TOK_TIMESTAMPTZ
     | KW_TIMESTAMP KW_WITH KW_LOCAL KW_TIME KW_ZONE -> TOK_TIMESTAMPLOCALTZ
     //| KW_TIMESTAMP KW_WITH KW_TIME KW_ZONE -> TOK_TIMESTAMPTZ
```
Comment (deniskuzZ) [inline parser/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g:2442]: should we drop this as well?

File: parser/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g [hunk 4]
```diff
@@ -263,11 +263,8 @@ TOK_TABLEROWFORMATLINES;
 TOK_TABLEROWFORMATNULL;
 TOK_TABLEFILEFORMAT;
 TOK_FILEFORMAT_GENERIC;
-TOK_OFFLINE;
 TOK_ENABLE;
 TOK_DISABLE;
-TOK_READONLY;
```
Comment (deniskuzZ) [inline parser/src/java/org/apache/hadoop/hive/ql/parse/HiveParser.g:269]: why is that removed? is it legacy and not used anymore?

**Review by deniskuzZ — APPROVED** [review]
LGTM +1

**Comment by okumin** [comment]
Thanks
